libs/community/langchain_community/utilities/nvidia_riva/common.py

```diff
@@ -0,0 +1,195 @@
+"""A collection of common code used by the NVIDIA Riva Runnables."""
+import asyncio
+import queue
+from enum import Enum
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    AsyncGenerator,
+    AsyncIterator,
+    Dict,
+    Generic,
+    Iterator,
+    Optional,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from langchain.pydantic_v1 import (
+    AnyHttpUrl,
+    BaseModel,
+    Field,
+    parse_obj_as,
+    root_validator,
+    validator,
+)
+from langchain_core.runnables import RunnableConfig, RunnableSerializable
+
+if TYPE_CHECKING:
+    import riva.client
+
+
+def _import_riva_client() -> "riva.client":
+    """Import the riva client and raise an error on failure."""
+    try:
+        # pylint: disable-next=import-outside-toplevel # this client library is optional
+        import riva.client
+    except ImportError as err:
+        raise ImportError(
+            "Could not import the NVIDIA Riva client library. "
+            "Please install it with `pip install nvidia-riva-client`."
+        ) from err
+    return riva.client
+
+
+# Create a generic Sentinel type
+class SentinelT:  # pylint: disable=too-few-public-methods
+    """An empty Sentinel type."""
+
+
+_TRANSFORM_END = SentinelT()
+
+
+# Define base Riva data types
+_InputT = TypeVar("_InputT")
+_OutputT = TypeVar("_OutputT")
+
+
+# pylint: disable-next=too-few-public-methods
+class RivaBase(Generic[_InputT, _OutputT], RunnableSerializable[_InputT, _OutputT]):
+    """A common set of methods for all Riva Runnables."""
+
+    @root_validator(pre=True)
+    @classmethod
+    def _validate_environment(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """Validate the Python environment and input arguments."""
+        _ = _import_riva_client()
+        return values
+
+    async def atransform(
+        self,
+        input: AsyncIterator[_InputT],
+        _: Optional[RunnableConfig] = None,
+    ) -> AsyncGenerator[_OutputT, None]:
+        """Intercept async transforms and route them to the synchronous transform in a thread."""
+        loop = asyncio.get_running_loop()
+        input_queue: queue.Queue[Union[_InputT, SentinelT]] = queue.Queue()
+        out_queue: asyncio.Queue[Union[_OutputT, SentinelT]] = asyncio.Queue()
+
+        async def _producer() -> None:
+            """Produce input into the input queue."""
+            async for val in input:
+                input_queue.put(val)
+            input_queue.put(_TRANSFORM_END)
+
+        def _consumer() -> None:
+            """Consume the input with transform."""
+            input_iterator = cast(Iterator[_InputT], iter(input_queue.get, _TRANSFORM_END))
+            for val in self.transform(input_iterator):
+                out_queue.put_nowait(val)
+            out_queue.put_nowait(_TRANSFORM_END)
+
+        async def _consumer_coro() -> None:
+            """Coroutine that wraps the consumer."""
+            await loop.run_in_executor(None, _consumer)
+
+        loop.create_task(_producer())
```

Comment:
I think these have to be awaited at the end, otherwise might be GCed before done

Reply:
Good eye. Adding this now.